### PR TITLE
Implement searchSortedLower Hint

### PR DESCRIPTION
**Pull request by joseSalazar4**
@cicr99 @rodrigo-pino @MaksymMalicki  Hey there! This is a draft of what I have so far, there's no tests yet due to  me not sure if the go implementation is okay. This tries to implement [this issue](https://github.com/NethermindEth/cairo-vm-go/issues/253)

### CODE REVIEW

File: pkg/hintrunner/zero/zerohint_memcpy.go [hunk 1]
```diff
@@ -28,3 +33,98 @@ func createVMExitScopeHinter(resolver hintReferenceResolver) (hinter.Hinter, err
...
+			indexAddr, err := index.GetAddress(vm)
+			if err != nil {
+				return err
+			}
+			//> for i in range(n_elms):
+			//> 	if memory[array_ptr + elm_size * i] >= ids.key:
+			//> 		ids.index = i
+			//> 		break
+			for i := uint64(0); i < n_elms.Uint64(); i++ {
+				v := memory.MemoryValueFromFieldElement(utils.Felt127.Add(array_ptr, elm_size))
```
Comment (MaksymMalicki) [inline pkg/hintrunner/zero/zerohint_memcpy.go:114]: I belive, you have to resolve array_ptr as an address, add elem_size * i using ```array_ptr.AddOffset()```. Then resolve the read value as felt and compare it with key

Comment (joseSalazar4) [reply pkg/hintrunner/zero/zerohint_memcpy.go:114]: Thanks for the hand, makes lot of sense that method. Changed!

**Comment by joseSalazar4** [comment]
I see that there's no test file for memcpy so is it necessary for this issue to be solved? 🤔 😄 Anyways, comments addressed

File: pkg/hintrunner/zero/zerohint_memcpy.go [hunk 2]
```diff
@@ -28,3 +33,98 @@ func createVMExitScopeHinter(resolver hintReferenceResolver) (hinter.Hinter, err
...
+	return newSearchSortedLowerHint(array_ptr, elm_size, n_elms, key, index), nil
+}
+
+func newSearchSortedLowerHint(array_ptr, elm_size, n_elms, key, index hinter.ResOperander) hinter.Hinter {
+	return &GenericZeroHinter{
+		Name: "SearchSortedLower",
+		Op: func(vm *VM.VirtualMachine, _ *hinter.HintRunnerContext) error {
+
+			//> array_ptr = ids.array_ptr
+			array_ptr, err := hinter.ResolveAsFelt(vm, array_ptr)
```
[the remark this answers is not in the document]
Comment (joseSalazar4) [reply pkg/hintrunner/zero/zerohint_memcpy.go]: Modified!

File: pkg/hintrunner/zero/zerohint_memcpy.go [hunk 3]
```diff
@@ -28,3 +33,98 @@ func createVMExitScopeHinter(resolver hintReferenceResolver) (hinter.Hinter, err
...
+			if err != nil {
+				return err
+			}
+
+			//> elm_size = ids.elm_size
+			elm_size, err := hinter.ResolveAsFelt(vm, elm_size)
+
+			//> assert isinstance(elm_size, int) and elm_size > 0, \
+			//> 	f'Invalid value for elm_size. Got: {elm_size}.'
+			if err != nil && utils.FeltLt(&utils.FeltZero, elm_size) {
```
[the remark this answers is not in the document]
Comment (joseSalazar4) [reply pkg/hintrunner/zero/zerohint_memcpy.go]: Modified! I used the isZero() func

File: pkg/hintrunner/zero/zerohint_memcpy.go [hunk 4]
```diff
@@ -28,3 +33,98 @@ func createVMExitScopeHinter(resolver hintReferenceResolver) (hinter.Hinter, err
...
+				return fmt.Errorf("Invalid value for elm_size. Got: %v.", elm_size)
+			}
+
+			//> n_elms = ids.n_elms
+			//> assert isinstance(n_elms, int) and n_elms >= 0, \
+			//> 	f'Invalid value for n_elms. Got: {n_elms}.'
+			n_elms, err := hinter.ResolveAsFelt(vm, n_elms)
+			if err != nil && utils.FeltLt(&utils.FeltZero, n_elms) {
```
[the remark this answers is not in the document]
Comment (joseSalazar4) [reply pkg/hintrunner/zero/zerohint_memcpy.go]: Modified!

File: pkg/hintrunner/zero/zerohint_memcpy.go [hunk 5]
```diff
@@ -28,3 +33,98 @@ func createVMExitScopeHinter(resolver hintReferenceResolver) (hinter.Hinter, err
...
+				return fmt.Errorf("Invalid value for n_elms. Got: %v.", n_elms)
+			}
+
+			var __find_element_max_size *fp.Element
+			// TODO: How to do this --> if __find_element_max_size is a variable?
```
[the remark this answers is not in the document]
Comment (joseSalazar4) [reply pkg/hintrunner/zero/zerohint_memcpy.go]: I am using the context to get the var now. Modified!

File: pkg/hintrunner/zero/zerohint_memcpy.go [hunk 6]
```diff
@@ -28,3 +33,119 @@ func createVMExitScopeHinter(resolver hintReferenceResolver) (hinter.Hinter, err
...
+
+			//> n_elms = ids.n_elms
+			//> assert isinstance(n_elms, int) and n_elms >= 0, \
+			//> 	f'Invalid value for n_elms. Got: {n_elms}.'
+			n_elms, err := hinter.ResolveAsFelt(vm, n_elms)
+
+			if err != nil {
+				return err
+			}
+			if n_elms.IsZero() {
```
Comment (cicr99) [inline pkg/hintrunner/zero/zerohint_memcpy.go:91]: See here that the condition is `n_elms >= 0`, hence the fact that `n_elms.IsZero()` is true does not contradict the previous fact. I would suggest using `FeltLt` utility instead

File: pkg/hintrunner/zero/zerohint_memcpy.go [hunk 7]
```diff
@@ -28,3 +33,119 @@ func createVMExitScopeHinter(resolver hintReferenceResolver) (hinter.Hinter, err
...
+			}
+
+			//> elm_size = ids.elm_size
+			elm_size, err := hinter.ResolveAsFelt(vm, elm_size)
+			if err != nil {
+				return err
+			}
+			//> assert isinstance(elm_size, int) and elm_size > 0, \
+			//> 	f'Invalid value for elm_size. Got: {elm_size}.'
+			if elm_size.IsZero() {
```
Comment (cicr99) [inline pkg/hintrunner/zero/zerohint_memcpy.go:79]: Again here, just to be safe, I'd use `FeltLe` with the zero value instead (we have the felt for zero in the utils/contants file you can use)

File: pkg/hintrunner/zero/zerohint_memcpy.go [hunk 8]
```diff
@@ -28,3 +33,119 @@ func createVMExitScopeHinter(resolver hintReferenceResolver) (hinter.Hinter, err
...
+
+			elementMaxSize, err := ctx.ScopeManager.GetVariableValue("__find_element_max_size")
+			if err != nil {
+				return err
+			}
+
+			elementMaxSizeFelt, ok := elementMaxSize.(*fp.Element)
+			if !ok {
+				return err
+			}
```
Comment (cicr99) [inline pkg/hintrunner/zero/zerohint_memcpy.go:103]: Just for a better understanding of the code, could you move the comment below this lines up? I feel that this logic belongs to that part as well

File: pkg/hintrunner/zero/zerohint_memcpy.go [hunk 9]
```diff
@@ -28,3 +33,119 @@ func createVMExitScopeHinter(resolver hintReferenceResolver) (hinter.Hinter, err
...
+
+			elementMaxSize, err := ctx.ScopeManager.GetVariableValue("__find_element_max_size")
+			if err != nil {
+				return err
+			}
+
+			elementMaxSizeFelt, ok := elementMaxSize.(*fp.Element)
+			if !ok {
+				return err
```
Comment (cicr99) [inline pkg/hintrunner/zero/zerohint_memcpy.go:102]: You should return a custom error here, something like "expected __find_element_max_size to be a felt".
Notice that at this point in the code `err` is `nil` from line 96, so you wouldn't actually know why your program failed

File: pkg/hintrunner/zero/zerohint_memcpy.go [hunk 10]
```diff
@@ -28,3 +33,119 @@ func createVMExitScopeHinter(resolver hintReferenceResolver) (hinter.Hinter, err
...
+			}
+
+			indexAddr, err := index.GetAddress(vm)
+			if err != nil {
+				return err
+			}
+
+			//> for i in range(n_elms):
+			for i := uint64(0); i < n_elms.Uint64(); i++ {
```
Comment (cicr99) [inline pkg/hintrunner/zero/zerohint_memcpy.go:125]: `n_elems` is a felt so in theory it could be as big as to have 252 bits, but it doesn't make sense for it to not fit in uint64, or we wouldn't be able to compute it. So it's ok if we cast it to uint64, but I would like you to do it separately and return an error if it's not possible to do so.

File: pkg/hintrunner/zero/zerohint_memcpy.go [hunk 11]
```diff
@@ -28,3 +33,119 @@ func createVMExitScopeHinter(resolver hintReferenceResolver) (hinter.Hinter, err
...
+	return newSearchSortedLowerHint(array_ptr, elm_size, n_elms, key, index), nil
+}
+
+func newSearchSortedLowerHint(array_ptr, elm_size, n_elms, key, index hinter.ResOperander) hinter.Hinter {
+	return &GenericZeroHinter{
+		Name: "SearchSortedLower",
+		Op: func(vm *VM.VirtualMachine, ctx *hinter.HintRunnerContext) error {
+
+			//> array_ptr = ids.array_ptr
+			array_ptr, err := hinter.ResolveAsAddress(vm, array_ptr)
```
Comment (cicr99) [inline pkg/hintrunner/zero/zerohint_memcpy.go:67]: Can we stick to camelCase for variable names?
You'll have to fix this and the rest of the variables to use the same format

File: pkg/hintrunner/zero/zerohint_memcpy.go [hunk 12]
```diff
@@ -28,3 +33,119 @@ func createVMExitScopeHinter(resolver hintReferenceResolver) (hinter.Hinter, err
...
+			}
+
+			indexAddr, err := index.GetAddress(vm)
+			if err != nil {
+				return err
+			}
+
+			//> for i in range(n_elms):
+			for i := uint64(0); i < n_elms.Uint64(); i++ {
+				array_ptr.AddOffset(int16(elm_size.Uint64()) * int16(i))
```
Comment (cicr99) [inline pkg/hintrunner/zero/zerohint_memcpy.go:126]: Notice that this operation modifies the value of `array_ptr`, which means for the next iteration the result of this operation won't be correct anymore. I see 2 approaches here:
1- create a newArrayPtr each time with the result of this operation without modifying arrayPtr at all.
2- You can modify arrayPtr by adding `elm_size` in each iteration. this should yield the correct result each time. Notice here the first iteration i = 0, so nothing is added to arrayPtr. I would update this variable at the end of the loop.

I personally like the second approach better so you don't have to create extra variables.

File: pkg/hintrunner/zero/zerohint_memcpy.go [hunk 13]
```diff
@@ -28,3 +33,119 @@ func createVMExitScopeHinter(resolver hintReferenceResolver) (hinter.Hinter, err
...
+				arrayPtrValue := memory.MemoryValueFromMemoryAddress(array_ptr)
+				arrayValue, err := arrayPtrValue.FieldElement()
+
+				if err != nil {
+					return err
+				}
+
+				fe := arrayValue
+				//> 	if memory[array_ptr + elm_size * i] >= ids.key:
+				if utils.FeltLt(key, fe) {
```
Comment (cicr99) [inline pkg/hintrunner/zero/zerohint_memcpy.go:136]: You can just use `arrayValue` directly, no need to save it in `fe` variable

Comment (cicr99) [inline pkg/hintrunner/zero/zerohint_memcpy.go:136]: the condition is `memory[array_ptr + elm_size * i] >= ids.key` so it's actually `FeltLe` the function you need to use

File: pkg/hintrunner/zero/zerohint_memcpy.go [hunk 14]
```diff
@@ -28,3 +33,119 @@ func createVMExitScopeHinter(resolver hintReferenceResolver) (hinter.Hinter, err
...
+					//> 		ids.index = i
+					//> 		break
+					indexValue := memory.MemoryValueFromFieldElement((new(fp.Element).SetInt64(int64(i))))
```
Comment (cicr99) [inline pkg/hintrunner/zero/zerohint_memcpy.go:139]: You could use `SetUint64` directly so you don't have to make the conversion for `i` where you do `SetInt64(int64(i))`

**Review by cicr99 — CHANGES_REQUESTED** [review]

**Comment by TAdev0** [comment]
Hi @joseSalazar4 , you have unit tests failing, can you make them pass before a review?

File: pkg/hintrunner/zero/zerohint_memcpy_test.go [hunk 15]
```diff
@@ -21,6 +22,116 @@ func TestZeroHintMemcpy(t *testing.T) {
...
+				operanders: []*hintOperander{
+					{Name: "array.el0", Kind: apRelative, Value: feltInt64(0)},
+					{Name: "array.el1", Kind: apRelative, Value: feltInt64(1)},
+					{Name: "array.el2", Kind: apRelative, Value: feltInt64(2)},
+					{Name: "array.el3", Kind: apRelative, Value: feltInt64(3)},
+					{Name: "arrayPtr", Kind: apRelative, Value: addr(0)},
+					{Name: "elmSize", Kind: fpRelative, Value: feltInt64(4)},
+					{Name: "nElms", Kind: fpRelative, Value: feltInt64(9)},
+					{Name: "index", Kind: fpRelative, Value: feltInt64(3)},
+					{Name: "key", Kind: fpRelative, Value: feltInt64(0)},
```
Comment (cicr99) [inline pkg/hintrunner/zero/zerohint_memcpy_test.go:119]: I believe these values don't make a lot of sense.
- `nElms` should be the amount of elements in the array
- `index` should have been `uninitialized` kind, because this is something that will be assigned during the execution of the hint. In fact, instead of checking you get a nil error after running the test, you should check the value that was saved in `index` address. This variable will represent the index of the first element of the array that is bigger or equal to the value of `key`
- `key` should have the value you are looking for. You can play with this in the tests, for example if you have array [0, 1, 2, 3] as in this case, apart from trying simple cases, you can try something edge case like `key = 5`. Since all elements of the array are smaller then the else statement after the for should be executed getting `index = 4`.

A suggestion is to try an array with different numbers than the ones you are using so the values of the elements don't coincide with their actual indexes. Notice the array has to be sorted in increasing order

**Review by cicr99 — CHANGES_REQUESTED** [review]

File: pkg/hintrunner/zero/zerohint_memcpy_test.go [hunk 16]
```diff
@@ -21,6 +22,116 @@ func TestZeroHintMemcpy(t *testing.T) {
...
+				makeHinter: func(ctx *hintTestContext) hinter.Hinter {
+					return newSearchSortedLowerHint(
+						ctx.operanders["arrayPtr"],
+						ctx.operanders["elmSize"],
+						ctx.operanders["nElms"],
+						ctx.operanders["key"],
+						ctx.operanders["index"],
+					)
+				},
+				errCheck: errorTextContains("failed obtaining the variable: __find_element_max_size"),
```
[the remark this answers is not in the document]
Comment (joseSalazar4) [reply pkg/hintrunner/zero/zerohint_memcpy_test.go:83]: The code checks if the var exists on the scope it does not check the value so do you think this testcase is needed?

[the remark this answers is not in the document]
Comment (cicr99) [reply pkg/hintrunner/zero/zerohint_memcpy_test.go:82]: This specific case shouldn't be necessary, as if the value is not in scope then it shouldn't even fail. But just was letting you know that if you need to work with a specific value for that variable then you need to assign it in the scope through ctxInit field

**Comment by rodrigo-pino** [comment]
Hey @joseSalazar4 how is it going. Do you need help with anything?

**Comment by TAdev0** [comment]
hi @joseSalazar4  are you still working on this?
